asset/asset.go

```diff
@@ -757,42 +769,27 @@ func NewRawKeyGenesisSigner(priv *btcec.PrivateKey) *RawKeyGenesisSigner {
 // For minting the first asset in a group, only one Genesis object is
 // needed, since we tweak with and sign over the same Genesis object.
 // The final tweaked public key and the signature are returned.
-func (r *RawKeyGenesisSigner) SignGenesis(keyDesc keychain.KeyDescriptor,
-	initialGen Genesis, currentGen *Genesis) (*btcec.PublicKey,
+func (r *RawKeyGenesisSigner) SignVirtualTx(signDesc *lndclient.SignDescriptor,
+	virtualTx *wire.MsgTx, prevOut *wire.TxOut) (*btcec.PublicKey,
 	*schnorr.Signature, error) {
 
-	if !keyDesc.PubKey.IsEqual(r.privKey.PubKey()) {
+	signerPubKey := r.privKey.PubKey()
+
+	if !signDesc.KeyDesc.PubKey.IsEqual(signerPubKey) {
 		return nil, nil, fmt.Errorf("cannot sign with key")
 	}
 
-	// TODO(jhb): Update to two-phase tweak
-	tweakedPrivKey := txscript.TweakTaprootPrivKey(
-		*r.privKey, initialGen.GroupKeyTweak(),
-	)
-
-	// If the current genesis is not set, we are minting the first asset in
-	// the group. This means that we use the same Genesis object for both
-	// the key tweak and to create the asset ID we sign. If the current
-	// genesis is set, the asset type of the new asset must match the type
-	// of the first asset in the group.
-	id := initialGen.ID()
-	if currentGen != nil {
-		if initialGen.Type != currentGen.Type {
-			return nil, nil, fmt.Errorf("asset group type mismatch")
-		}
-
-		id = currentGen.ID()
-	}
+	sig, err := SignVirtualTx(r.privKey, signDesc, virtualTx, prevOut)
 
-	// TODO(roasbeef): this actually needs to sign the digest of the asset
-	// itself
-	idHash := sha256.Sum256(id[:])
-	sig, err := schnorr.Sign(tweakedPrivKey, idHash[:])
+	// Derive the final tweaked public key also used for signing.
+	tweakedGroupKey, err := GroupPubKey(
+		signerPubKey, signDesc.SingleTweak, signDesc.TapTweak,
```

Comment:
See above, I think we can keep the sign desc as is, then it just derives the tweaks from: the asset passed in, and the genesis to use for signing purposes. We can also tweak that slightly to be `GroupKey` or w/e (assuming it includes the key and the genesis information).

Reply:
This way the interface is a bit more opinionated.

Reply:
Pretty sure this is addressed.